onadata/apps/logger/data_migration/restore_backup.py

```diff
@@ -19,33 +31,82 @@ def copy_attrs(from_obj, to_obj, attrs):
 class BackupRestorer(object):
     def __init__(self, xform, version=None, restore_last=False):
         self.xform = xform
+        self.old_xml = self.xform.xml
         self.xform_backup = self._get_xform_backup(version, restore_last)
 
     def restore_xform_backup(self):
         self._restore_xform_from_backup()
+        self.migrate_surveys_to_prev_schema()
 
-        for survey in self.xform.instances.iterator():
-            try:
-                backup_survey = self.xform_backup.surveys.get(uuid=survey.uuid)
-            except BackupInstance.DoesNotExist:
-                self.migrate_survey_to_prev_schema(survey)
-            else:
-                self._restore_survey_from_backup(survey, backup_survey)
-
-    def migrate_survey_to_prev_schema(self, survey):
-        changes = self.xform_backup.changes
-        reversed_decisions = self.get_reversed_migration_decisions(changes)
+    def migrate_surveys_to_prev_schema(self):
+        changes_list = self._get_migration_changes_in_between(self.xform, self.xform_backup)
+        decisions = MigrationDecisioner.convert_changes_to_decisions(
+            self._merge_migration_changes(changes_list)
+        )
+        # XXX: Temporarily give form backup old xml so that xml comparison in
+        # data migration works as exected. Otherwise it would compare against
+        # the same xmls
+        self.xform_backup.xml = self.old_xml
```

Comment:
This comment is an immediate "what?". Without digging into data_migrator_factory, I have questions:
- if you written "**temporarily** give form backup old xml", and I assume it's an assignment below, why isn't it reassigned immediately after data_migrator_factory call?
- can't self.xform_backup be set properly in previous steps?
- or, assign to different variable so that you get the desired side effects without changing self.xform_backup?

Please double check this, I think you should be able to express you intent clearly without this confusion

Reply:
You are right, this made the code obfuscated and wtf-ish. Basically what it was, it was a hack to provide previous xml for data migration by using and xform_backup as an intermediate object. 

When I'm migrating the data, it needs updated xform and its previous instance - so that i can compare these two and then migrate the data according to the differences. So, at first I needed to update the form to the desired version, and then perform the data migration.  Thus, i needed to pass the previous xml (b/c after restoring the form xform.xml was equal to xform_backup.xml)